### PR TITLE
feat: show oneOf/anyOf as a dropdown

### PR DESCRIPTION
**Pull request by marcelltoth**
Resolves stoplightio/elements#667
Will resolve stoplightio/elements#853

First attempt at replacing `anyOf` / `oneOf` with a dropdown, developed together during a hacking session. 

Mostly for @P0lip to review and let us know if we are on the right track.

### CODE REVIEW

File: src/tree/tree.ts [hunk 1]
```diff
@@ -32,10 +33,27 @@ export class SchemaTreeListTree extends TreeListTree {
...
+  private _schemaToTreeMap!: WeakMap<SchemaNode, TreeListNode>;
+  private _treeToSchemaMap!: WeakMap<TreeListNode, SchemaNode>;
   private _flattenedSchemaNodes!: WeakSet<RegularNode>;
 
+  private _selectedIndex: number = 0;
+
+  @action
+  public updateSelectedTypeIndex = (treeListNode: TreeListNode, newIndex: number) => {
+    this._selectedIndex = newIndex;
+    this.replaceNode(treeListNode, this.buildTreeFragment((treeListNode.metadata as any).schemaNode.parent));
```
Comment (P0lip) [inline src/tree/tree.ts:45]: Urgh. For a long time I've been willing to fix the signature of `metadata`.
It's such a PITA, particularly in Studio.

File: src/tree/tree.ts [hunk 2]
```diff
@@ -118,32 +137,53 @@ export class SchemaTreeListTree extends TreeListTree {
...
-
     return treeNode;
   }
 
-  public buildTreeFragment(schemaNode: SchemaNode) {
-    const treeNode = this.createTreeNode(schemaNode);
+  public buildTreeFragment(rawSchemaNode: SchemaNode): SchemaTreeListNode {
+    let treeNode: SchemaTreeListNode = this.createTreeNode(rawSchemaNode);
+    const schemaNode =
+      isRegularNode(rawSchemaNode) && ['anyOf', 'oneOf'].includes(rawSchemaNode.combiners?.[0] ?? '')
```
Comment (P0lip) [inline src/tree/tree.ts:184]: Be careful.
`oneOf`/`anyOf` may undergo a merging process.
See https://github.com/stoplightio/json-schema-tree/blob/beta/src/walker/walker.ts#L287
I don't know how we want to tackle that behavior from a product perspective, but ti's certainly something to watch out for.
I suspect no changes are likely to be required in the piece of code I linked above, but you might want to verify that.

File: src/tree/tree.ts [hunk 3]
```diff
@@ -32,10 +33,27 @@ export class SchemaTreeListTree extends TreeListTree {
...
-  private _treeToSchemaMap!: WeakMap<SchemaTreeListNode, SchemaNode>;
+  private _schemaToTreeMap!: WeakMap<SchemaNode, TreeListNode>;
+  private _treeToSchemaMap!: WeakMap<TreeListNode, SchemaNode>;
   private _flattenedSchemaNodes!: WeakSet<RegularNode>;
 
+  private _selectedIndex: number = 0;
+
+  @action
+  public updateSelectedTypeIndex = (treeListNode: TreeListNode, newIndex: number) => {
+    this._selectedIndex = newIndex;
```
Comment (P0lip) [inline src/tree/tree.ts:44]: I suppose you'll want to maintain it per node basis.

File: src/components/shared/Property.tsx [hunk 4]
```diff
@@ -40,6 +40,20 @@ export const Property: React.FunctionComponent<IProperty> = ({ onGoToRef }) => {
...
 
+      {treeListNode.metadata?.typeOptions && (
+        <select
+          onChange={e => {
+            schemaTree.updateSelectedTypeIndex(treeListNode, Number(e.currentTarget.value));
+          }}
+          onClick={e => e.stopPropagation()}
+        >
+          {treeListNode.metadata.typeOptions.map((option: SchemaNode, index: number) => (
+            <option key={index} value={index}>
```
Comment (P0lip) [inline src/components/shared/Property.tsx:51]:
```suggestion
            <option key={option.id} value={index}>
```

**Review by P0lip — COMMENTED** [review]
Looking pretty good so far

**Comment by marcelltoth** [comment]
Closing this for now, but will get reopen as we start working on the main issue.

**Pull request closed without merging**